connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -544,6 +555,95 @@ public int generation() {
+
+        addRequest(
+            new Callable<Void>() {
+                @Override
+                public Void call() throws Exception {
+                    ConfigDef configDef = getConfig(connType);
+                    Collection<ConfigDef.ConfigKey> configKeys = configDef.configKeys().values();
+                    List<ConfigKeyInfo> configKeyInfos = new LinkedList<>();
+                    for (ConfigDef.ConfigKey configKey: configKeys) {
+                        configKeyInfos.add(convertConfigKey(configKey));
+                    }
+                    ConfigDefInfo configDefInfo = new ConfigDefInfo(connType, configKeyInfos);
+                    callback.onCompletion(null, configDefInfo);
+                    return null;
+                }
+            },
+            forwardErrorCallback(callback)
+        );
+    }
+
+    @Override
+    public synchronized void validateConfigs(final String connType, final Map<String, String> connectorConfig, final Callback<ConfigInfos> callback) {
+        log.trace("Submitting validate configuration request {}", connType);
+
+        addRequest(
+            new Callable<Void>() {
+                @Override
+                public Void call() throws Exception {
+                    ConnectorConfig connConfig = new ConnectorConfig(connectorConfig);
+                    String connName = connConfig.getString(ConnectorConfig.NAME_CONFIG);
+                    ConfigDef configDef = getConfig(connType);
+                    Connector connector = getConnector(connType, connName);
+                    List<Config> configs = connector.validate(configDef, connectorConfig);
+                    List<ConfigInfo> configInfoList = new LinkedList<>();
+                    int errorCount = 0;
+                    for (Config config : configs) {
+                        ConfigInfo configInfo = new ConfigInfo(config.getName(), config
+                            .getValue(), config.getRecommendedValues(), config.getErrorMessages(), config.isVisible());
+                        configInfoList.add(configInfo);
+                        errorCount += config.getErrorMessages().size();
+                    }
+                    ConfigInfos
+                        configInfos = new ConfigInfos(connName, errorCount, configInfoList);
+                    callback.onCompletion(null, configInfos);
+                    return null;
+                }
+            },
+            forwardErrorCallback(callback)
+        );
+
+    }
+
+    private ConfigDef getConfig(String connType) {
+        if (configs.containsKey(connType)) {
+            return configs.get(connType);
+        } else {
+            ConfigDef configDef = worker.getConnectorConfigDef(connType);
+            configs.put(connType, configDef);
+            return configDef;
+        }
+    }
+
+    private Connector getConnector(String connType, String connName) {
+        if (tempConnectors.containsKey(connName)) {
```

Comment:
What is the need for `tempConnectors`? Why should a temporary connector exist any longer than the configuration validation request? Why should we even need to track any state for that temporary connector? I would expect that two validation requests, even if concurrent and for the same connector ID, should be independent since it is just validating the config, not executing any connector code with it.